project_role/models/project_assignment.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2018 Brainbean Apps (https://brainbeanapps.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class ProjectAssignment(models.Model):
+    _name = 'project.assignment'
+    _description = 'Project Assignment'
+    _inherit = ['mail.thread', 'mail.activity.mixin']
+
+    name = fields.Char(
+        compute='_compute_name',
+        store=True,
+        index=True,
+    )
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        string='Company',
+        related='project_id.company_id',
+        store=True,
+        readonly=True,
+    )
+    project_id = fields.Many2one(
+        comodel_name='project.project',
+        string='Project',
+        required=True,
+        track_visibility='onchange',
+    )
+    role_id = fields.Many2one(
+        comodel_name='project.role',
+        string='Role',
+        required=True,
+        track_visibility='onchange',
+    )
+    user_id = fields.Many2one(
+        comodel_name='res.users',
+        string='User',
+        required=True,
+        track_visibility='onchange',
+    )
+
+    _sql_constraints = [
+        (
+            'project_role_user_uniq',
+            'unique (project_id, role_id, user_id)',
+            'User may be assigned per role only once within a project!'
+        ),
+    ]
+
+    @api.depends('project_id.name', 'role_id.name', 'user_id.name')
+    def _compute_name(self):
+        for assignment in self:
+            assignment.name = _('%s as %s on %s') % (
+                assignment.user_id.name,
+                assignment.role_id.name,
+                assignment.project_id.name,
+            )
+
+    @api.multi
+    @api.constrains('role_id', 'user_id')
+    def _check_assignable(self):
+        for assignment in self:
+            if not assignment.role_id.can_assign(assignment.user_id):
+                raise ValidationError(_(
+                    'User %s can not be assigned to role %s.'
+                ) % (
+                    assignment.user_id.name,
+                    assignment.role_id.name,
+                ))
+
+    @api.model
+    def create(self, values):
+        values = self._assignment_preprocess(values)
+        assignment = super().create(values)
+        assignment._assignment_postprocess(values)
+        return assignment
+
+    @api.multi
+    def write(self, values):
+        values = self._assignment_preprocess(values)
+        result = super().write(values)
+        self._assignment_postprocess(values)
+        return result
+
+    @api.multi
+    def unlink(self):
+        self._assignment_preremove()
+        values = self.read()
+        result = super().unlink()
+        self._assignment_postremove(values)
+        return result
+
+    def _assignment_preprocess(self, values):
```

Comment:
I don't think this is needed. You can inherit `write` and `create` as well in your extension module.

Reply:
It's not _needed_, yet similar approach has been used in https://github.com/odoo/odoo/blob/97aec9f836bf4154e95fa9857cef5e31cbad5d00/addons/hr_timesheet/models/hr_timesheet.py#L134 and I like to explicitly point to extension points that are by-design.

Reply:
Got your point, have to think about it a bit more. Also, @aheficent since you have some extension in mind, how this impacts your idea?